pkg/scheduler/queue/tree_queue.go

```diff
@@ -52,241 +145,133 @@ func (q *TreeQueue) IsEmpty() bool {
-// complete operations on the dequeued item, but failure is not yet final, and the
-// operations should be retried by a subsequent queue consumer.
-//
-// childPath must be relative to the receiver node; providing a QueuePath beginning with
-// the receiver/parent node name will create a child node of the same name as the parent.
-func (q *TreeQueue) EnqueueFrontByPath(childPath QueuePath, v any) error {
-	childQueue, err := q.getOrAddNode(childPath)
+func (n *Node) enqueueBackByPath(tree *TreeQueue, pathFromNode QueuePath, v any) error {
+	childNode, err := n.getOrAddNode(pathFromNode, tree)
 	if err != nil {
 		return err
 	}
-	if childQueue.localQueue == nil {
-		childQueue.localQueue = list.New()
-	}
-	childQueue.localQueue.PushFront(v)
+	childNode.localQueue.PushBack(v)
 	return nil
 }
 
-// getOrAddNode recursively adds tree queue nodes based on given relative child path.
-//
-// childPath must be relative to the receiver node; providing a QueuePath beginning with
-// the receiver/parent node name will create a child node of the same name as the parent.
-func (q *TreeQueue) getOrAddNode(childPath QueuePath) (*TreeQueue, error) {
-	if len(childPath) == 0 {
-		return q, nil
-	}
+func (n *Node) dequeue() (QueuePath, any) {
+	var v any
+	var childPath QueuePath
 
-	if q.childQueueMap == nil {
-		q.childQueueMap = make(map[string]*TreeQueue, 1)
-	}
+	path := QueuePath{n.name}
 
-	var childQueue *TreeQueue
-	var ok bool
-	if childQueue, ok = q.childQueueMap[childPath[0]]; !ok {
-		// no child node matches next path segment
-		// create next child before recurring
-		childQueue = NewTreeQueue(childPath[0])
+	if n.IsEmpty() {
+		return path, nil
+	}
 
-		// add new child queue to ordered list for round-robining;
-		// in order to maintain round-robin order as nodes are created and deleted,
-		// the new child queue should be inserted directly before the current child
-		// queue index, essentially placing the new node at the end of the line
-		if q.currentChildQueueIndex == localQueueIndex {
-			// special case; cannot slice into childQueueOrder with index -1
-			// place at end of slice, which is the last slot before the local queue slot
-			q.childQueueOrder = append(q.childQueueOrder, childQueue.name)
+	var checkedAllNodes bool
+	var dequeueNode *Node
+	// continue until we've found a value or checked all nodes that need checking
+	for v == nil && !checkedAllNodes {
+		dequeueNode, checkedAllNodes = n.dequeueAlgorithm.dequeueGetNode(n)
+		var deletedNode bool
+		// dequeuing from local queue
+		if dequeueNode == n {
+			if n.localQueue.Len() > 0 {
+				// dequeueNode is self, local queue non-empty
+				if elt := n.localQueue.Front(); elt != nil {
+					n.localQueue.Remove(elt)
+					v = elt.Value
+				}
+			}
+		} else if dequeueNode == nil {
+			// no dequeue-able child found; reset checked children to 0,
+			// as we won't update state before moving on
+			n.childrenChecked = 0
+			return path, v
```

Comment:
IMHO `childrenChecked` should be algorithm state, not tree state - all uses of it are in the algorithm, not the tree and we can just call to update the state here before returning.

It's the algo's job rather than the tree's job to a) decide if it's done selecting nodes and b) set itself to the correct state after a successful or failed dequeue.

Also there's some awkwardness with having `!checkedAllNodes` be a condition on the outerloop, but it never comes into play there because we will never return to the top of the loop with a case where where v is nil `checkedAllNodes` is true - because we early return here.
We should do one or the other - don't early return here, or don't check that flag in the outer loop

Reply:
I considered keeping `childrenChecked` in the state, but decided against it in the end. With `tenantQuerierAssignments`, we explicitly take advantage of things in state being _shared_ across all nodes at the same level. Things like `childrenChecked` (and, e.g., `queuePosition`) apply to/are maintained by individual nodes, not shared state, so it seemed counterintuitive and possibly error-prone to put them in algorithm state.